### PR TITLE
display association rules plots in streamlit

### PR DESCRIPTION
**Pull request by batmanscode**
# issue #884

### Changes

Added `display_format`

```
def plot_model(model, plot="2d", scale=1, display_format=None):
.
.
.
    if display_format=='streamlit'
        st.write(fig)
    else:
        fig.show()

```

### Notes

* Didn't include `import streamlit as st` since this will only work inside streamlit anyway
* Didn't use [`st.plotly_chart`](https://docs.streamlit.io/en/stable/api.html#streamlit.plotly_chart) since [`st.write`](https://docs.streamlit.io/en/stable/api.html#streamlit.write) can plot many types of charts and won't have to be changed if plotly is dropped later

# Update

* Added `display_format` to anywhere `plot_model` can be called 😃
* Error handling

### Additional:
In [nlp.py](https://github.com/pycaret/pycaret/blob/plot-in-streamlit/pycaret/nlp.py) where `fig.iplot()` is used, I had to change [`asFigure = save_param`](https://github.com/pycaret/pycaret/blob/2d528733269dcb353fe8d88f0c90f02c6bc8c593/pycaret/nlp.py#L2101) to `asFigure = True` because streamlit needs a plotly object to interpret (https://discuss.streamlit.io/t/cufflinks-in-streamlit/2232/2)

```
.
.
.
if display_format=='streamlit':
    fig = df.iplot(asFigure=True) # plotly obj needs to be returned for streamlit to interpret
    st.write(fig)
else:
    df.iplot()
```

### CODE REVIEW

**Comment by Yard1** [comment]
Hey @batmanscode, thanks for your contribution! I am not familiar with streamlit, but can it be applied to other plots too - for example, in regression or clustering?

**Comment by batmanscode** [comment]
Hi @Yard1, yes absolutely! Adding a conditional `st.write(fig)` anywhere there is a `fig.show()` will work.

I would like to add `display_format='streamlit'` to the plots in other modules as well but it seems for some the change will have to be in `..internal/tabular` and that will take a little more time for me to go through.

And as a side note, I do recommend checking out streamlit when you have the time 😃

**Comment by batmanscode** [comment]
Okay so maybe `..internal/tabular` doesn't really need it since there is an option to save the plot in `plot_model()` so a streamlit app can just display the saved image.

**Comment by Yard1** [comment]
I think it would be a good idea to have a unified API. Do you think you could add that? If you have trouble with modifying `internal/tabular`, I am happy to help.

**Comment by batmanscode** [comment]
Yeah I think you're right. Makes more sense to have it uniform.

Yes, I should be able to! Thank you, I will let you know if I run into any problems 😃

**Comment by batmanscode** [comment]
Hi @Yard1, my last commit is failing some tests but I'm not sure why. Can you please have a look when you have some time?

**Comment by batmanscode** [comment]
Btw, for plots like on [this](https://github.com/pycaret/pycaret/blob/2d528733269dcb353fe8d88f0c90f02c6bc8c593/pycaret/nlp.py#L2092) line:

image

Why is it `df3 = df2.iplot()`? I thought it should be `df2.iplot()` or `iplot(df3)`

**Comment by batmanscode** [comment]
> Btw, for plots like on [this](https://github.com/pycaret/pycaret/blob/2d528733269dcb353fe8d88f0c90f02c6bc8c593/pycaret/nlp.py#L2092) line:
> 
> image
> 
> Why is it `df3 = df2.iplot()`? Based on some tutorials I saw, I thought it should be `df2.iplot()` or `iplot(df3)`. Or have I missed something? 😅

I wonder if this could be related to #319

**Comment by Yard1** [comment]
@batmanscode Apologies, I must have missed this. I'll take a look as soon as I am able.

**Comment by batmanscode** [comment]
No problem at all @Yard1, I just figured you must have been busy.

**Comment by Yard1** [comment]
@batmanscode Could you merge master into this branch, so the tests can pass? Thanks.

**Comment by Yard1** [comment]
@batmanscode So the issue was that `evaluate_models` needs to be updated any time there's a new parameter added to `plot_models`. One more thing before I merge this, though - could the default value for `display_format` be `"matplotlib"`, and could you make it throw an exception if neither `"matplotlib"` nor `"streamlit"` is passed as the `display_format` parameter? It's easier to understand than just `None`. Thanks.

**Comment by batmanscode** [comment]
Oh right, thank you for the fix and explanation 😃 will keep that in mind for any future changes.

Okay sure, that does sound clearer. I will go through and add this exception. And I think you mean `"plotly"` 😆

I do have a concern about naming the default plots:
Right now `plot_model` uses plotly for every module and in this case `plotly` being default for `display_format` makes sense but, if in a future release, some modules use something other than plotly, it might get a little weird.

Or do you think that it's unlikely and `plot_model` will always use one plotting library in every module?

**For example:**
If the nlp module uses some other library for plotting, then its `plot_model` would have `display_format=newlib` or `streamlit` while all the other modules would have `display_format=plotly` or `streamlit` which would mean all the `display_format` options won't work universally.

In this case, another option could be to change `display_format` to something like `special_formatting=None` or `streamlit`. This way each module can have its own default plotting library without causing any confusion.

**Comment by Yard1** [comment]
> For example:
If the nlp module uses some other library for plotting, then its plot_model would have display_format=newlib or streamlit while all the other modules would have display_format=plotly or streamlit which would mean all the display_format options won't work universally.

> In this case, another option could be to change display_format to something like special_formatting=None or streamlit. This way each module can have its own default plotting library without causing any confusion.

Fair point. Let's keep `None` for now, then. Just please make sure that a user cannot pass an invalid value.

**Comment by batmanscode** [comment]
Alright will do. I think the if/else statements need to be changed as well. Right now things will work fine even if an invalid value is passed since it will just execute `fig.show()` if anything other than `'streamlit'` is given.

Right now it's:

```
if display_format == 'streamlit':
    st.write(fig)
else:
    fig.show()
```

But something like this is probably better:

```
if display_format == 'streamlit':
    st.write(fig)
if disploy_format == None:
    fig.show()
```

**Comment by Yard1** [comment]
@batmanscode It's fine to leave it as is, considering we'll have value checks beforehand.

**Comment by batmanscode** [comment]
Ah yes, thanks!

**Comment by batmanscode** [comment]
All done @Yard1. Sorry, forgot to comment that I was finished.

File: pycaret/internal/tabular.py [hunk 1]
```diff
@@ -5786,6 +5790,12 @@ def is_tree(e):
         raise TypeError(
             "feature parameter must be string containing column name of dataset."
         )
+        
+    # checking display_format parameter
+    plot_formats = [None, "streamlit"]
+    
+    if display_format not in plot_formats:
+        raise ValueError("display_format can only be None or \"streamlit\".")
```
Comment (Yard1) [inline pycaret/internal/tabular.py:5798]: Please use single quotes `'` instead of escaping double quotes

**Review by Yard1 — CHANGES_REQUESTED** [review]
Just one thing, please change this in all strings

File: pycaret/internal/tabular.py (hunk 1 above)
Comment (batmanscode) [reply pycaret/internal/tabular.py]: Done. Not sure why it's not passing all the checks though.

File: pycaret/arules.py [hunk 2]
```diff
@@ -247,11 +247,38 @@ def plot_model(
...
+    
+    # check if model is a pandas dataframe
+    if isinstance(model, pd.DataFrame) == False:
+        raise TypeError("Model needs to be a pandas.DataFrame object.")
+        
+    # check plot parameter
+    plot_types = ["2d", "3d"]
+    
+    if plot not in plot_types:
+        raise ValueError("Plots can only be \"2d\" or \"3d\".")
```
Comment (Yard1) [inline pycaret/arules.py:270]: Quotes

**Review by Yard1 — CHANGES_REQUESTED** [review]
One last thing - can you run black on all modified files? Thanks.

File: pycaret/arules.py (hunk 2 above)
Comment (batmanscode) [reply pycaret/arules.py]: Sorry, I missed this.

**Comment by batmanscode** [comment]
> One last thing - can you run black on all modified files? Thanks.

Done 👍🏻

**Comment by Yard1** [comment]
Thanks, looks great.

**Comment by Yard1** [comment]
@batmanscode Hey, are you on PyCaret slack? Could you join if you haven't already and DM me there? I have no experience with streamlit but would like to expand the support for it to yellowbrick visualizers and I could use your help. Thanks.

**Comment by batmanscode** [comment]
Thank you @Yard1, I've learnt a lot doing this 😃

Sure, I can join and contact you there, happy to help. I actually haven't used slack in a few years lol.

**Comment by nataliaasaa** [comment]
Hey! Does anyone knows a way to show evaluate_model results on streamlit? Thanksss

**Comment by batmanscode** [comment]
> Hey! Does anyone knows a way to show evaluate_model results on streamlit? Thanksss

Hi, `evaluate_model()` creates a little UI in a notebook using ipywidgets (I think) so this will not work in a streamlit app, at least right now

You can create your own `evaluate_model()` like function/script to work in streamlit using `plot_model()` (which is what `evaluate_model()` uses internally as well)

Something like this should work 😊:

```
"""
# Evaluate Model
"""

plots = st.selectbox('select plot', ("auc", "threshold")) # docs will have the other plots, not all are supported

plot_model(model, plot=plots, display_format="streamlit")
```

**Pull request merged**